tests/lib/ocpp/v16/test_smart_charging_handler.cpp

```diff
@@ -0,0 +1,764 @@
+#include <filesystem>
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+namespace fs = std::filesystem;
+
+#include <database_handler_mock.hpp>
+#include <evse_security_mock.hpp>
+#include <ocpp/common/call_types.hpp>
+#include <ocpp/v16/smart_charging.hpp>
+#include <optional>
+
+
+namespace ocpp {
+namespace v16 {
+
+/**
+ * Chargepoint Test Fixture
+ *
+ * Test Matrix:
+ *
+ * Positive Boundary Conditions:
+ * - PB01 Valid Profile
+ * - PB02 Valid Profile No startSchedule & handler allows no startSchedule & profile.chargingProfileKind == Absolute
+ * - PB03 Valid Profile No startSchedule & handler allows no startSchedule & profile.chargingProfileKind == Relative
+ * - PB04 Absolute ChargePointMaxProfile Profile with connector id 0
+ * - PB05 Absolute TxDefaultProfile
+ * - PB06 Absolute TxProfile ignore_no_transaction == true
+ * - PB07 Absolute TxProfile && connector transaction != nullptr && transaction_id matches SKIPPED: was not able to test
+ *
+ * Negative Boundary Conditions:
+ * - NB01 Valid Profile, ConnectorID gt this->connectors.size()
+ * - NB02 Valid Profile, ConnectorID lt 0
+ * - NB03 profile.stackLevel lt 0
+ * - NB04 profile.stackLevel gt profile_max_stack_level
+ * - NB05 profile.chargingProfileKind == Absolute && !profile.chargingSchedule.startSchedule
+ * - NB06 Number of installed Profiles is > max_charging_profiles_installed
+ * - NB07 Invalid ChargingSchedule
+ * - NB08 profile.chargingProfileKind == Recurring && !profile.recurrencyKind
+ * - NB09 profile.chargingProfileKind == Recurring && !startSchedule
+ * - NB10 profile.chargingProfileKind == Recurring && !startSchedule && !allow_charging_profile_without_start_schedule
+ * - NB11 Absolute ChargePointMaxProfile Profile with connector id not 0
+ * - NB12 Absolute TxProfile connector_id == 0
+ */
+class ChargepointTestFixture : public testing::Test {
+protected:
+    void SetUp() override {
+    }
+
+    // Adding transactiosn to the connectors because Profiles other than ChargePointMaxProfile need it in order to be valid.
+    void addConnector(int id) {
```

Comment:
It looks like these first few functions are utility functions that are used by the tests - I would suggest moving them to a `test_smart_charging_handler_utils.cpp` file to ensure that the test file only has tests.

I have not used `gtest` in particular before, so if this is the recommended practice for it, please add a link to the source of the recommendation.

Reply:
Done. Moved ChargepointTestFixture to its own file.